### PR TITLE
phase 2: mattes/migrate -> migratex

### PR DESCRIPTION
**Pull request by rdallman**
* changes format of migrations to migratex format
* updates test runner to use new interface (double checked this with printlns,
the tests go fully down and then up, and work on pg/mysql)
* fix some migratex bugs thanks to ^
* add context to migratex.Migration api

largely a follow up to #787 (see discussion there, if curious) but w/o having to migrate migrators

### CODE REVIEW

File: api/datastore/sql/migrations/README.md [hunk 1]
```diff
@@ -2,39 +2,43 @@
 
 All migration files should be of the format:
 
-`[0-9]+_[add|remove]_model[_field]*.[up|down].sql`
+`[0-9]+_[add|remove]_model[_field]*.sql`
 
 The number at the beginning of the file name should be monotonically
 increasing, from the last highest file number in this directory. E.g. if there
-is `11_add_foo_bar.up.sql`, your new file should be `12_add_bar_baz.up.sql`.
+is `11_add_foo_bar.sql`, your new file should be `12_add_bar_baz.sql`.
```
Comment (denismakogon) [inline api/datastore/sql/migrations/README.md]: This should be *.go file

File: api/datastore/sql/migrations/README.md [hunk 2]
```diff
@@ -2,39 +2,43 @@
...
+	"github.com/pressly/goose"
+)
+
+
+func up1(tx *sql.Tx) error {
+	_, err := tx.Exec("...")
+	return err
+}
+
+func down1(tx *sql.Tx) error {
```
Comment (denismakogon) [inline api/datastore/sql/migrations/README.md]: would be the conflict, should be `up8`, `down8`

File: api/datastore/sql/migrations/README.md [hunk 3]
```diff
@@ -2,39 +2,43 @@
...
+	_, err := tx.Exec("...")
+	return err
+}
+
+func init() {
+	migrations = append(migrations, &goose.Migration{
+		Version:    int64(8),
+		UpFn:       up1,
+		DownFn:     down1,
+		Registered: true,
```
Comment (denismakogon) [inline api/datastore/sql/migrations/README.md]: this needs to be updated, this migration has goose references.

File: api/datastore/sql/migrations/README.md [hunk 4]
```diff
@@ -2,39 +2,43 @@
...
-may refer to https://github.com/mattes/migrate/blob/master/MIGRATIONS.md which
-illustrates some of the finer points.
-
-After creating the file you will need to run, in the same directory as this
-README:
-
-```sh
-$ go generate
 ```
+Each migration should have initialize `goose.Migration` struct with corresponding version, up/down function, filename (source field).
```
Comment (denismakogon) [inline api/datastore/sql/migrations/README.md]: the same

**Comment by rdallman** [comment]
thanks, forgot to update it! will do

**Comment by rdallman** [comment]
@denismakogon updated

**Comment by rdallman** [comment]
woo, ready to land https://github.com/fnproject/fn/pull/641 !

**Pull request merged**